File: vcc/CodeModel2VccHelper/Token.cs
//-----------------------------------------------------------------------------
//
// Copyright (C) Microsoft Corporation.  All Rights Reserved.
//
//-----------------------------------------------------------------------------

namespace Microsoft.Research.Vcc
{
  public abstract class Token
  {
    abstract public int Line { get; }
    abstract public int Column { get; }
    abstract public int Byte { get; }
    abstract public string Filename { get; }
    abstract public string Value { get; }

    public virtual bool SuppressWarning(int code)
    {
      return false;
    }

    public static Token NoToken
    {
      get { return DummyToken.Instance; }
    }

    public Token Related { get; protected set; }
  }

  public class DummyToken : Token
  {
    static DummyToken inst;

    private DummyToken() { }

    public static DummyToken Instance
    {
      get
      {
        if (inst == null) inst = new DummyToken();
        return inst;
      }
    }

    public override int Line
    {
      get { return 0; }
    }

    public override int Column
    {
      get { return 0; }
    }

    public override int Byte
    {
      get { return 0; }
    }

    public override string Filename
    {
      get { return "no_file"; }
    }

    public override string Value
    {
      get { return ""; }
    }
  }

  public class LazyToken : Token
  {

    public delegate Token GetTokenDelegate();

    private readonly GetTokenDelegate getToken;

    public LazyToken(GetTokenDelegate getToken) {
      this.getToken = getToken;
    }

    private Token token;
    public Token DelayedToken {
      get {
        if (this.token == null) {
          this.token = this.getToken();
        }
        return this.token;
      }
    }

    public override int Line {
      get { return this.DelayedToken.Line; }
    }

    public override int Column {
      get { return this.DelayedToken.Column; }
    }

    public override int Byte {
      get { return this.DelayedToken.Byte; }
    }

    public override string Filename {
      get { return this.DelayedToken.Filename; }
    }

    public override string Value {
      get { return this.DelayedToken.Value; }
    }
  }
}

File: vcc/Ast/Token.fs
namespace Microsoft.Research.Vcc

[<AbstractClass>]
type public Token() =
  
  let mutable related : Token = Token.NoToken

  abstract Line : int with get
  abstract Column : int with get
  abstract Byte : int with get
  abstract Filename : string with get
  abstract Value : string with get

  abstract member SuppressWarning : int->bool
  default this.SuppressWarning _ = false

  member this.Related
    with get() = related
    and set(value) = related <- value

  static member NoToken = DummyToken.Instance

and public DummyToken() =
  inherit Token()

  static member private instance = new DummyToken() :> Token

  override this.Line = 0
  override this.Column = 0
  override this.Byte = 0
  override this.Filename = "<no file>"
  override this.Value = System.String.Empty

  static member Instance with get() = DummyToken.instance

and public LazyToken(getToken : unit -> Token) =
  inherit Token()

  let token = lazy getToken()

  override this.Line = token.Value.Line
  override this.Column = token.Value.Column
  override this.Byte = token.Value.Byte
  override this.Filename = token.Value.Filename
  override this.Value = token.Value.Value
  override this.SuppressWarning(code) = token.Value.SuppressWarning(code)

